static/usr/lib/core/rtc-sys-time-init

```diff
@@ -0,0 +1,112 @@
+    esac
+
+    # Apply priority level to all lines
+    printf "%s\n" "$@" | while IFS= read -r line; do
+        printf "%s\n" "$lvl$line"
+    done
+}
+
+# Expects single argument RTC device node /dev/rtc* (instance name dev-rtc* with escaping undone)
+dev_node="/$1"
+
+# Get systemd timestamp file created during core build
+clock_epoch_stamp=0
+if [ -e "/usr/lib/clock-epoch" ]; then
+    clock_epoch_stamp="$(stat -L "/usr/lib/clock-epoch" -c '%Y')"
+else
+    journal warning "Warning: cannot stat /usr/lib/clock-epoch"
+fi
+clock_epoch_date=$(date -d @"$clock_epoch_stamp")
+journal debug "core/systemd build timetamp: $clock_epoch_date"
+
+# Get systemd-timesyncd timestamp file initialized to systemd build time and updated periodically
+# (default 60s) and at shutdown
+timesync_clock_stamp=0
+if [ -e "/var/lib/systemd/timesync/clock" ]; then
+    timesync_clock_stamp="$(stat -L "/var/lib/systemd/timesync/clock" -c '%Y')"
+else
+    journal warning "Warning: cannot stat /var/lib/systemd/timesync/clock"
+fi
+timesync_clock_date=$(date -d @"$timesync_clock_stamp")
+journal debug "timesyncd timestamp: $timesync_clock_date"
+
+most_recent_stamp=$(echo "$clock_epoch_stamp $timesync_clock_stamp" | tr ' ' '\n' | sort -rn | head -1)
+
+rtc_attr_hctosys=0
+sys_time_set_res=1
+if udevadm info --attribute-walk --name="$dev_node" | grep -q 'ATTR{hctosys}=="1"'; then
+    rtc_attr_hctosys=1
+    sys_time_set_res=0
+    echo "Kernel synced system time to RTC $dev_node time"
+else
+    echo "Reading RTC $dev_node time..."
+    if ! rtc_stamp=$(hwclock -r -u -f "$dev_node" 2>/dev/null); then
+        journal error "Error: cannot read RTC $dev_node"
+    else
+        rtc_stamp=$(date -d "$rtc_stamp" +%s)
+        rtc_date=$(date -d @"$rtc_stamp")
+        echo "Successfully read RTC time: $rtc_date"
+
+        if [ "$rtc_stamp" -gt "$most_recent_stamp" ]; then
+            echo "RTC time is valid, syncing system time..."
+            if hwclock -s --noadjfile -u -f "$dev_node" 2>/dev/null; then
+                sys_time_set_res=0
+                echo "Successfully synced system time"
+            else
+                journal error "Error: cannot sync system time"
+            fi
+        else
+            journal error "Error: RTC time invalid, skipping system time sync"
+        fi
+    fi
+fi
+
+# Check if system time was successfully synced to RTC time
+if [ "$sys_time_set_res" -ne 0 ] || [ "$rtc_attr_hctosys" -eq 1 ]; then
+    # Move system time forward if required
+    now_stamp="$(date +'%s')"
+    if [ "$now_stamp" -lt "$most_recent_stamp" ]; then
+        echo "Moving system time forward to most recent timestamp..."
+        if date -s @"$most_recent_stamp" >/dev/null 2>&1; then
+            echo "Successfully moved system time forward "
+        else
+            sys_time_set_res=1
+            journal error "Error: cannot forward system time"
+        fi
+    else
+        echo "System time is ahead of most recent timestamp, skipping update"
```

Comment:
I would not say "skipping update". This one is a fixup. I would call say "skipping fixup".

Reply:
Improved